internal/audit/audit.go

```diff
@@ -119,177 +138,124 @@ func Batch(ctx context.Context, secrets []string, secStore secretGetter, expirat
 
 	bar := termio.NewProgressBar(int64(len(secrets)))
 	bar.Hidden = ctxutil.IsHidden(ctx)
-
-	i := 0
-	for secret := range checked {
-		if secret.err != nil {
-			en := secret.err.Error()
-			errors[en] = append(errors[en], secret.name)
-		} else if secret.content != "" {
-			duplicates[secret.content] = append(duplicates[secret.content], secret.name)
-		}
-		for _, m := range secret.messages {
-			messages[m] = append(messages[m], secret.name)
-		}
-
+	a.pcb = func() {
 		bar.Inc()
-		i++
-		if i == len(secrets) {
-			break
-		}
 	}
-	bar.Done()
 
-	return auditPrintResults(ctx, duplicates, messages, errors)
-}
+	for i := 0; i < maxJobs; i++ {
+		<-done
+	}
+	bar.Done()
 
-func audit(ctx context.Context, secStore secretGetter, validators []validator, expiry time.Duration, secrets <-chan string, checked chan<- auditedSecret, done chan struct{}) {
-	if expiry < time.Hour {
-		expiry = DefaultExpiration
+	if err := a.checkHIBP(ctx); err != nil {
+		return nil, err
 	}
 
+	return a.r.Finalize(), nil
+}
+
+func (a *Auditor) audit(ctx context.Context, secrets <-chan string, done chan struct{}) {
 	for secret := range secrets {
-		as := auditedSecret{
-			name: secret,
-		}
 		// check for context cancelation.
 		select {
 		case <-ctx.Done():
-			as.err = errors.New("user aborted")
-			checked <- as
-
 			continue
 		default:
 		}
 
-		debug.Log("Checking %q", secret)
-
-		// handle old passwords
-		revs, err := secStore.ListRevisions(ctx, secret)
-		if err != nil {
-			as.messages = append(as.messages, err.Error())
-		} else if len(revs) > 0 && time.Since(revs[0].Date) > expiry {
-			as.messages = append(as.messages, fmt.Sprintf("Password too old (%dd)", int(expiry.Hours()/24)))
-		}
-
-		if len(validators) < 1 {
-			checked <- as
-
-			continue
-		}
-
-		sec, err := secStore.Get(ctx, secret)
-		if err != nil {
-			debug.Log("Failed to check %s: %s", secret, err)
-			as.err = err
-			if sec != nil {
-				as.content = sec.Password()
-			}
-			// failed to properly retrieve the secret.
-			checked <- as
-
-			continue
-		}
-		as.content = sec.Password()
+		a.auditSecret(ctx, secret)
+	}
+	done <- struct{}{}
+}
 
-		// do not check empty secrets.
-		if as.content == "" {
-			checked <- as
+func (a *Auditor) auditSecret(ctx context.Context, secret string) {
+	debug.Log("Auditing %q", secret)
 
-			continue
-		}
+	// handle old passwords
+	revs, err := a.s.ListRevisions(ctx, secret)
+	if err != nil {
+		a.r.AddError(secret, err)
+	}
+	if len(revs) > 0 {
+		a.r.SetAge(secret, time.Since(revs[0].Date))
+	}
 
-		// handle password validation errors.
-		if errs := allValid(validators, secret, sec); len(errs) > 0 {
-			for _, e := range errs {
-				as.messages = append(as.messages, e.Error())
-			}
-			checked <- as
+	sec, err := a.s.Get(ctx, secret)
+	if err != nil {
+		debug.Log("Failed to check %s: %s", secret, err)
 
-			continue
+		a.r.AddError(secret, err)
+		if sec != nil {
+			a.r.AddPassword(secret, sec.Password())
 		}
```

Comment:
In any case it's missing the check `sec.Password()` is not empty unlike done below.

Reply:
No. If we get an error but still a non-nil secret (not sure how that could happen), we still want to record the password.

Reply:
You're doing a 
```
        // do not check empty secrets.
	if sec.Password() == "" {
		return
	}
	
	// add the password for the duplicate check
	a.r.AddPassword(secret, sec.Password())
``` 
below but not here, so in case of empty password (e.g. an entry used only for its keys) and of an error, it might be incorrectly flagged.

Reply:
I think you are right, but I agree that it's not a high priority right now.
The reports, even after this PR, are still not great. I want to improve them, but there is already a lot in this PR and at some point I should stop breaking things ;)